lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -9386,83 +9386,112 @@ namespace Js
+                    // But first we need to give iterator a chance to clean up.
+
+                    prop = JavascriptOperators::GetProperty(iterator, PropertyIds::return_, scriptContext);
+                    if (prop != iterator->GetLibrary()->GetUndefined())
                     {
-                        JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedObject);
+                        RecyclableObject* method = RecyclableObject::FromVar(prop);
+                        Var args[] = { iterator, yieldData->data };
+                        CallInfo callInfo(CallFlags_Value, _countof(args));
+                        Var result = JavascriptFunction::CallFunction<true>(method, method->GetEntryPoint(), Arguments(callInfo, args));
+
+                        if (!JavascriptOperators::IsObject(result))
+                        {
+                            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedObject);
+                        }
                     }
 
-                    Var value = JavascriptOperators::GetProperty(RecyclableObject::FromVar(result), PropertyIds::value, scriptContext);
-                    // CONSIDER: Using an exception to carry the return value and force finally code to execute is a bit of a janky
-                    // solution since we have to override the value here in the case of yield* expressions.  It works but is there
-                    // a more elegant way?
-                    //
-                    // Instead what if ResumeYield was a "set Dst then optionally branch" opcode, that could also throw? Then we could
-                    // avoid using a special exception entirely with byte code something like this:
-                    //
-                    // ;; Ry is the yieldData
-                    //
-                    // ResumeYield Rx Ry $returnPathLabel
-                    // ... code like normal
-                    // $returnPathLabel:
-                    // Ld_A R0 Rx
-                    // Br $exitFinallyAndReturn
-                    //
-                    // This would probably give better performance for the common case of calling next() on generators since we wouldn't
-                    // have to wrap the call to the generator code in a try catch.
-                    yieldData->exceptionObj->SetThrownObject(value);
+                    // Use ThrowExceptionObject() to get debugger support for breaking on throw
+                    JavascriptExceptionOperators::ThrowExceptionObject(yieldData->exceptionObj, yieldData->exceptionObj->GetScriptContext(), true);
+                }
+                else if (!isNext)
+                {
+                    // Do not use ThrowExceptionObject for return() API exceptions since these exceptions are not real exceptions
+                    throw yieldData->exceptionObj;
                 }
             }
-            else if (!isReturn)
+
+            RecyclableObject* method = RecyclableObject::FromVar(prop);
```

Comment:
It is not safe to call `RecyclableObject::FromVar` on a random `Var`.  The `Var` could be a tagged int or float.  `RecyclableObject::Is` should be used first to check that it is indeed a `RecyclableObject`.  If it isn't that would be equivalent to the spec's `GetMethod` line that checks if the result is callable, so a TypeError should be thrown if it isn't.  There is a helper, `JavascriptConversion::IsCallable` that you could use to check for this and make the code read nicer than `RecyclableObject::Is`.